src/binaryen-shell.cpp

```diff
@@ -53,16 +53,67 @@ struct ExitException {
 //
 
 struct ShellExternalInterface : ModuleInstance::ExternalInterface {
-  char *memory;
+  // The underlying memory can be accessed through unaligned pointers which
+  // isn't well-behaved in C++. WebAssembly nonetheless expects it to behave
+  // properly. Avoid emitting unaligned load/store by checking for alignment
+  // explicitly, and performing memcpy if unaligned.
+  class Memory {
+    // Use char because it doesn't run afoul of aliasing rules.
+    std::vector<char> memory;
+    template <typename T>
+    bool aligned(size_t address) {
+      static_assert(!(alignof(T) & (alignof(T) - 1)), "must be a power of 2");
+      return 0 == (address & (alignof(T) - 1));
+    }
+    Memory(Memory&) = delete;
+    Memory& operator=(const Memory&) = delete;
+
+   public:
+    Memory() {}
+    void resize(size_t newSize) {
+      // Allocate at least this size to get proper alignment.
+      // The code is optimistic this will work until wg21.link/p0035r0 happens.
+      const size_t minSize = 1 << 12;
```

Comment:
1. But for that, you only need 7 extra bytes. The minimum size here is `1 << 12 = 4096`?
2. Why does it need to be aligned in the first place? All the methods here are designed to work properly even if it isn't?